### ISSUE TITLE
Refresh token not functioning properly, forcing daily login - (Error: invalid_grant: expired access/refresh token)

### ISSUE THREAD

**Issue by scottpelak** [joeferraro/MavensMate#747]
MavensMate v0.0.11-beta.4 doesn't appear to save Project Credentials.  I love the new interface and love the use of Salesforce Authentication instead of Security Tokens.  MM App doesn't appear to store Project Credentials.  Idk if this is limited due to use of SF Auth.  Is it possible for the SF Auth Page to remember the Username / Password similar to an HTML Page?

Additionally to Username / Password, MM App v0.0.11-beta.4 doesn't remember how to login to the Project, meaning it doesn't remember if the Project Type is:
- Production
- Sandbox
- Developer
- Pre-Release

- **MavensMate App Version**:
  MavensMate App v0.0.11-beta.4

- **Editor**:
  Sublime

- **Platform/Version**:
  Windows 10 64-bit

- **Behind firewall/proxy**:
  None

- **Salesforce Server(s)**:
  CS61

Open App.  Open Project.  Login to project by entering Credentials.  Close App.  Open Project.  Will have to re-enter Credentials.

**Comment by joeferraro** [comment]
@scottpelak do you have keychain/keyring enabled in Desktop settings?

**Comment by daeschwed** [comment]
maybe this additional info helps you out, I experience the same behavior, though this can be worked around by re-authenticating, so I can live with it.

my setup:
OS X El Capitan
Version 10.11.6

Sublime Text 3
Stable Channel, Build 3114

MM App v0.0.11-beta.5
MavensMate for Sublime Text v7.0.1

Behind firewall/proxy: None
Salesforce Server(s): EU4, CS83 (guess this doesn't apply anyway…)

Trying to edit an existing project throws an error about authenticating.
I have «Use keyring (mm_use_keyring)» enabled. Within Keychain Access, I find 3 sets of records clearly originating from MM APP:
- 3x MavensMate-accessToken
- 3x MavensMate-refreshToken

Still, re-opening a project throws an authentication error.
Here's the last part from mavensmate.log on MM_LOG_LEVEL «DEBUG»

`
2016-09-20T07:42:49.267Z - error:    invalid_grant: expired access/refresh token
    at /Applications/MavensMate.app/Contents/Resources/app/node_modules/jsforce/lib/oauth2.js:190:19
    at tryCallOne (/Applications/MavensMate.app/Contents/Resources/app/node_modules/promise/lib/core.js:37:12)
    at /Applications/MavensMate.app/Contents/Resources/app/node_modules/promise/lib/core.js:123:15
    at flush (/Applications/MavensMate.app/Contents/Resources/app/node_modules/asap/raw.js:50:29)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)
2016-09-20T07:42:49.267Z - debug:   project has expired access/refresh token, marking as invalid
2016-09-20T07:42:49.267Z - error:   Could not initiate existing Project instance: expired access/refresh token
2016-09-20T07:42:49.267Z - warn:    Project requiring re-auth added to client
2016-09-20T07:42:49.268Z - info:    Project added to client, but it requires authentication, redirecting to auth endpoint
2016-09-20T07:42:49.268Z - error:   requested route/resource requires re-authentication, sending 500
`


my current workaround is to re-authenticate each day.
Not as bad as it sounds.
Drop me a line if you need more detailed informations.

Keep up the brilliant work, MM App looks very promising - love the Lightning style it comes with.

Cheers

**Comment by joeferraro** [comment]
@daeschwed are you by chance sharing a login with other developers? There is a known, undocumented limit on the number of active grants per user.

**Comment by daeschwed** [comment]
@joeferraro nope, those are personal logins.
same happens if a try with a Developer org and no deployment connections

**Comment by joeferraro** [comment]
@daeschwed and how do you use Desktop? Do you keep it open throughout the day or are you continually starting/exiting it throughout the day. Just trying to understand whether perhaps several token requests over a relatively short period of time is somehow causing the refreshToken to expire prematurely.

**Comment by harshal-gajare** [comment]
@joeferraro  I think I am seeing a similar behaviour as well where I have to re-authenticate each day for the same project.  I usually have my MM-app open all through the day and I believe if I close it the next day for any reason and reopen it , I get an error.  I will again tomorrow and post logs if I run into the same error or at least provide an update.

**Comment by ChuckJonas** [comment]
Seeing the same thing... Have to login every day.  Message: `expired access/refresh token` Really wish you had left in the option to use Password+Token.

**Comment by daeschwed** [comment]
@joeferraro same as Chuck and Harshal, I usually have it open throughout working days.

@ChuckJonas have you tried disabling mm_use_keyring? just be sure not to commit any credentials to your repo.
haven't tried it myself, though.

I believe the password+token thingie relies on TLS 1.0, so that train left in June, at least for Sandbox instances.
And last boarding call for Production instances will be in march 2017

**Comment by harshal-gajare** [comment]
@joeferraro I tried this morning and I get the error 

> MavensMate is unable to connect to Salesforce. Please re-authenticate to continue.
> This can happen if your access/refresh token have been revoked or if MavensMate is unable to retrieve your credentials from secure storage.

**Comment by ChuckJonas** [comment]
@daeschwed fairly certain that TLS 1.0 retirement doesn't have any impact on password + token login (other than the fact that you can't use TLS 1.0 to authenticate).   I've used Mavensmate (with user/pass+token) on plenty of orgs that have that security update turned on. 

It's possible that there is some other update coming down the pipe that I'm unaware of that has made @joeferraro decided to move to oAuth...

**Comment by andreyzh** [comment]
I've got the same issue with latest beta (v5).
If I keep MM Desktop open throughout the day and don't close it when PC enters sleep mode, it's working fine. If the desktop is closed, next day I have to re-authenticate.
I am the only one using that account. Keyring is enabled.

**Comment by arkos** [comment]
I am also affected by this issues. Not sure if it's relevant, but the logs indicate that loginURL is login.salesforce.com while I am trying to log into a sandbox and not production.

**Comment by ChuckJonas** [comment]
@joeferraro:  @arkos is correct.  Just did some debugging and it appears that `conn.oauth2` is defaulting to `login.salesforce.com` oAuth URL's when they should actually be `test.salesforce.com` (`sfdc-client.js` |  ` _refreshFn()`).   

As a quick, test I manually overrode these variables and the refresh did in fact succeed! 

```
  logger.debug('manually refreshing access token...');
  //override to test refresh
  conn.oauth2.loginUrl = 'https://test.salesforce.com';
  conn.oauth2.authzServiceUrl = 'https://test.salesforce.com/services/oauth2/authorize';
  conn.oauth2.tokenServiceUrl = 'https://test.salesforce.com/services/oauth2/token';
  conn.oauth2.revokeServiceUrl = 'https://test.salesforce.com/services/oauth2/revoke';
  // danger: private jsforce api
  conn.oauth2._postParams({
```

I'll try to submit a PR out this evening.

**Comment by sinanbunni** [comment]
Has the issue being resolved. I am still facing it.

**Comment by daeschwed** [comment]
@sinanbunni the issue has been resolved with this PR.
what issue are you facing?
could you provide some details to your setup? what version are you running?
one thought - I am running MavensMate locally and recently noticed, that node v7 (current stable) doesn't play nicely with keytar, that handles secure storage.
might only apply if you too are running MavensMate through the terminal.

see initial post on this issue as a reference what to post about your setup.

Cheers

**Comment by jacobraccuia** [comment]
Hi I'm having authentication problems.  I have to constantly authenticate against a custom domain URL.


    Production
    Sandbox
    Developer
    Pre-Release

    MavensMate App Version:
    MavensMate App v0.0.11-beta.6

    Editor:
    Sublime

    Platform/Version:
    OSX 10.11.6

    Behind firewall/proxy:
    None

    Salesforce Server(s):
    CS62


I am not sure what else to provide but it is realllllly inconvenient to reauthenticate each day because I have to copy and paste my custom domain etc in constantly.  I will provide any information I need to help debug this!

**Comment by daeschwed** [comment]
@jacobraccuia inconvenient it is, but I'm sure we can get this figured out.

- do you have keychain enabled?
- are the logs showing anything interesting? (might wanna set logging level to DEBUG)
- also, in your projects root, check the file .settings in the config folder. Is the "instanceUrl" correctly set to your custom domain? this is the instance that MavensMate will try to authenticate to.

I just tested this with a project using a custom domain, and it works as expected.

What I did:
- create new project from MavensMate using custom domain
- in Salesforce setup, go to Security -> Session Management and remove any active sessions, particularly the ones with session type OAuth2
- reopen project in MavensMate -> this will force a token refresh event

now check mavensmate.log for any pointers as to why this fails.

Cheers

**Comment by harshal-gajare** [comment]
@jacobraccuia Please try with v0.0.11-beta.7. That has fixed the issue with the need to reauthenticate everyday.

**Comment by jacobraccuia** [comment]
@harshal-gajare I didn't realize that I was on an older version.  I have installed the latest version and will let you know if the project persists.

@daeschwed I would have replied to you sooner if I had realized there were responses on here.  I  didn't get an email notifying me of replies for some reason :(  Anyways thanks so much for your help, and I will follow the instructions you have laid out if the latest version doesn't resolve my problems.
